src/builders/polylines.js

```diff
@@ -40,579 +43,444 @@ export function buildPolylines (
+        texcoord_normalize = texcoord_normalize || DEFAULT.TEXCOORD_NORMALIZE;
+        texcoord_ratio = texcoord_ratio || DEFAULT.TEXCOORD_RATIO;
         var [min_u, min_v, max_u, max_v] = texcoord_scale || default_uvs;
     }
 
+    var v_scale = 1 / (width * texcoord_ratio * v_scale_adjust); // scales line texture as a ratio of the line's width
+
     // Values that are constant for each line and are passed to helper functions
     var context = {
         vertex_data,
         vertex_template,
-        half_width: width/2,
-        vertices: [],
+        half_width: width / 2,
         scaling_index,
         scaling_normalize,
-        scalingVecs: scaling_index && [],
         texcoord_index,
-        texcoords: texcoord_index && [],
         texcoord_normalize,
         min_u, min_v, max_u, max_v,
-        v_scale: 1 / ((width * texcoord_ratio) * v_scale_adjust), // scales line texture as a ratio of the line's width
-        total_dist: 0,
-        num_pairs: 0
+        miter_len_sq
     };
 
-    for (var ln = 0; ln < lines.length; ln++) {
-        // Remove dupe points from lines
-        var line = dedupeLine(lines[ln], closed_polygon);
-        if (!line) {
-            continue; // skip if no valid line remaining
-        }
+    var coordCurr, coordNext, normPrev, normNext;
+    var v = 0;
 
-        var lineSize = line.length;
+    for (var index = 0; index < lines.length; index++) {
+        var line = lines[index];
 
-        // Ignore non-lines
-        if (lineSize < 2) {
+        // Skip if line is not valid
+        if (line.length < 2) {
             continue;
         }
 
-        //  Initialize variables
-        var coordPrev = [0, 0], // Previous point coordinates
-            coordCurr = [0, 0], // Current point coordinates
-            coordNext = [0, 0]; // Next point coordinates
-
-        var normPrev = [0, 0],  // Right normal to segment between previous and current m_points
-            normCurr = [0, 0],  // Right normal at current point, scaled for miter joint
-            normNext = [0, 0];  // Right normal to segment between current and next m_points
-
-        var isPrev = false,
-            isNext = true;
-
-        // Add vertices to buffer according to their index
-        indexPairs(context);
-
-        // Do this with the rest (except the last one)
-        for (let i = 0; i < lineSize ; i++) {
-
-            // There is a next one?
-            isNext = i+1 < lineSize;
+        // Loop backwards through line to a tile boundary if found
+        if (closed_polygon && join_type === JOIN_TYPE.MITER) {
+            var boundaryIndex = getTileBoundaryIndex(line);
+
+            if (boundaryIndex !== 0) {
+                // create new line that is a cyclic permutation of the original
+                var newLine = line.splice(boundaryIndex);
+                Array.prototype.push.apply(newLine, line.slice(1));
+                newLine[newLine.length] = newLine[0];
+                lines.push(newLine);
```

Comment:
Separate but related issue: I'm seeing cases, such as with building feature `-3696564` (which you can isolate with `filter: { id: -3696564 }`) where new lines that are only a single element long (e.g. not a valid line) are being inserted.